db/src/Pos/DB/Functions.hs

```diff
@@ -7,91 +7,40 @@ module Pos.DB.Functions
        (
        -- * Encoded putting/getting
          dbGetBi
-       , dbGetBiNoVersion
        , dbPutBi
-       , dbPutBiNoVersion
-       , dbSerializeValue
 
        -- * Decoding/encoding primitives and iteration related
-       , dbDecode
        , encodeWithKeyPrefix
        , processIterEntry
        ) where
 
 import           Universum
 
 import qualified Data.ByteString as BS (drop, isPrefixOf)
-import           Formatting (bprint, builder, sformat, shown, stext, string,
-                     (%))
+import           Formatting (sformat, shown, string, (%))
 
 import           Pos.Binary.Class (Bi, decodeFull', serialize')
-import           Pos.Core.Configuration (HasCoreConfiguration,
-                     dbSerializeVersion)
 import           Pos.DB.Class (DBIteratorClass (..), DBTag, IterType,
                      MonadDB (..), MonadDBRead (..))
 import           Pos.DB.Error (DBError (..))
 import           Pos.Util.Util (maybeThrow)
 
--- | Read serialized value associated with given key from pure DB.
-dbGetBiNoVersion
-    :: forall v m.
-       (Bi v, MonadDBRead m)
-    => DBTag -> ByteString -> m (Maybe v)
-dbGetBiNoVersion tag key = do
-    bytes <- dbGet tag key
-    traverse (dbDecode . (ToDecodeValue key)) bytes
-
--- | Write serializable value to DB for given key.
-dbPutBiNoVersion :: (Bi v, MonadDB m) => DBTag -> ByteString -> v -> m ()
-dbPutBiNoVersion tag k v = dbPut tag k (serialize' v)
-
 -- | Read serialized value (with version) associated with given key from pure DB.
 dbGetBi
     :: forall v m.
        (Bi v, MonadDBRead m)
     => DBTag -> ByteString -> m (Maybe v)
-dbGetBi tag key = do
-    bytes <- dbGet tag key
-    val <- traverse (dbDecode . (ToDecodeValue key)) bytes
-    traverse onVersionError val
-  where
-    onVersionError :: (Word8, v) -> m v
-    onVersionError (verTag, v)
-        | verTag /= dbSerializeVersion =
-              throwM $ DBUnexpectedVersionTag dbSerializeVersion verTag
-        | otherwise = pure v
+dbGetBi tag key =
+    dbGet tag key >>= traverse (either throwM pure . dbDecodeIgnoreVersion)
 
 -- | Write serializable value to DB for given key. Uses simple versioning.
 dbPutBi :: (Bi v, MonadDB m) => DBTag -> ByteString -> v -> m ()
-dbPutBi tag k v = dbPut tag k (dbSerializeValue v)
-
--- | Version of 'serialize'' function that includes version when serializing a value.
-dbSerializeValue :: (HasCoreConfiguration, Bi a) => a -> ByteString
-dbSerializeValue = serialize' . (dbSerializeVersion,)
+dbPutBi tag k v = dbPut tag k (serialize' v)
```

Comment:
Hmm, want to check that a little more carefully on Monday.

Marking this "requires changes" until I properly examine the consequences of this.

Reply:
The weird thing is that from looking at the code, I expect this to break when accessing existing DBs. The fact that it doesn't freaks me out even more than if it did break.

I know its a pain in the neck, but we can we separate out the `dbSerializeVersion` parts of this PR and maybe even have a test to show that its safe. The idea would be to:
* Raise a PR with some tests that use `dbSerializeVersion` as it is now.
* Once that PR is merged, raise a PR removing `dbSerializeVersion`.

Reply:
You're right, that's safer and will reduce the complexity of this PR. I'm getting on it now.

Reply:
This changes the DB to write without a version, but read using:
```
dbDecodeIgnoreVersion :: forall v . Bi v => ByteString -> Either DBError v
dbDecodeIgnoreVersion bytes = case decodeFull' @v bytes of
    Right val -> Right val
    Left _    -> bimap DBMalformed snd $ decodeFull' @(Word8, v) bytes
```
which first tries to decode without a version and, if that fails, tries again with the version and drops it

Reply:
That's why it doesn't break on old DBs :)

Reply:
The tests are in place and passing. Ready to re-review, paying particular attention to [this file](https://github.com/input-output-hk/cardano-sl/pull/3550/files#diff-1f5ccc5e83ce78d1e0ac185490714322)